Add pod images selectors

cdoron opened the pull request:
**Affected functionality**
Two additional selectors for the Kubernetes Workload Attestor

**Description of change**
My code adds a new selector, called _pod-images_, which lists the container-images of all running containers within a pod (sorted alphabetically).
One use case for which this selector would be useful is that of Istio. A typical istio pod has two containers, one of which is an envoy proxy. It is the envoy proxy that establishes mTLS connections. When envoy sends a whoami query to spire (SDS), the response is “you are envoy”. This response may not be particularly helpful, since it does not uniquely identify the workload. It would be much more helpful to identify the workload as “envoy running in a pod alongside (and on the behalf of) the XXX container”.

Another selector I have added, called _pod-init-images_, lists the images of all containers in the InitContainerStatuses field.

My proposed implementation is straightforward. In getSelectorsFromPodInfo() I call a new method, getPodImages(), which goes over pod.Status.ContainerStatuses, and collects the image IDs of all running containers. These image IDs are then sorted and joined by comma. This returned string is then added to the new _pod-images_ selector. Similarly, the method collects the image IDs of all init container, sorts them and joins them by comma. This returned string is then added to the new _pod-init-images_ selector

evan2645 commented:
Thank you very much for this contribution @cdoron!

Implementation-wise, it all looks pretty good to me. The use case for the selector is also clear.

One thing I worry about as I review, however, is that the size of the value seems unbounded. Is there a limit to how many containers can be in a pod? I had a quick look at the datastore plugin, and we currently store selector value as `varchar(255)`. If we truncated in this case, we could get unlucky?

One idea that comes to mind is we could hash the list, but that certainly makes things a lot more cumbersome.

Curious to hear your thoughts here

cdoron commented:
Thanks @evan2645 : You make a very good point.
I will give it some more thought and make adjustments.

cdoron commented:
Hi @evan2645 : Your idea of using a hash seems to be the most reasonable.
I made the changes accordingly. Thanks

azdagron commented:
I'm a little concerned about the operator experience here. When I go to deploy a workload, I've got to run the same algorithm to generate the hash. This seems a little cumbersome. Even if we provide tooling there seems to be non-trivial room for error. Additionally, debugging a hash difference seems problematic. If all I see in the logs is that workload attestation produced hash X, and my entry has hash Y, how do I reconcile the two? It will take 1) knowing what input was used to produce the entry hash and 2) having a snapshot of the container images running in the pod at the time the log was generated.

That being said, I think the use case is **very compelling**. I'd like to see us support this somehow. I'm just not yet convinced a hash is going to be ergonomic.

One option is to use the original approach but enforce selector length at the API layer instead of the persistence layer. MySQL 5.7+ are in strict mode by default and will fail an entry being created with too long a selector, but unfortunately it becomes silent truncation in 5.5 and 5.6 (which we still support). Adding length enforcement (which we should probably do anyway) at least allows for use case, albeit under restricted circumstances (i.e. limits to how many containers you need to run in the pod).

I'm still musing on this.... like I said, supported something like this would be great and I'm really glad you brought it up @cdoron.

mcpherrinm commented:
If you could select on `k8s:pod-image:foo AND k8s:pod-image:envoy` would that suffice?  Notably, if there's some third container, say `baz`,  that would change the proposed `k8s:pod-images`, but would just add a third selector so `k8s:pod-image:foo AND k8s:pod-image:envoy` would continue matching.

If we added some support to SPIRE, we could plausibly have some way of expressing `k8s:pod-containers` has exactly `foo` and `envoy`.  That's definitely a bigger change, though, and would probably bring a whole lot of baggage of "what other matching operations might we want".

cdoron commented:
Hi Andrew, hi Matthew,
@azdagron : Matthew's suggestion eliminates hash computations and also addresses your debugging concern.
@mcpherrinm : We have added counts for number of containers in pod, and of number of containers of each type. This would allow a user to mandate that an identity is to given to a pod with `foo`and `envoy`, but no `baz` (or any other non-specified image).
Thanks!

azdagron commented on pkg/agent/plugin/workloadattestor/k8s/k8s.go:
```diff
@@ -603,6 +623,14 @@ func getSelectorsFromPodInfo(pod *corev1.Pod, status *corev1.ContainerStatus) []
 		makeSelector("pod-name:%s", pod.Name),
 		makeSelector("container-name:%s", status.Name),
 		makeSelector("container-image:%s", status.Image),
+		makeSelector("pod-container-count:%s", podContainerCount),
+		makeSelector("pod-init-container-count:%s", podInitContainerCount),
```
nit: getPodImages doesn't need to return the count since we can just use the length of the returned images
```suggestion
		makeSelector("pod-container-count:%s", len(podImages)),
		makeSelector("pod-init-container-count:%s", len(podInitImages)),
```

mcpherrinm commented:
```
pod-image:foo
pod-image:bar
pod-image-count:2
```
I really like this.  Can't believe we didn't think of using a count earlier 🤦

I wonder if the `count` could be a built-in feature somehow, which may be a future improvement.

cdoron commented:
Hi,
@azdagron, @mcpherrinm : As you suggested, I removed the counter for each container image, and left only the counter for distinct container images. I also changed the getPodImages method to only return a single value.
The previous version tried to distinguish between identities of two pods with the same set of container images, but different counts (for instance, one pod with one `foo` and one `bar` container, and another with one `foo` and two `bar`s). This is a niche scenario, and not really needed for our use case.
Thanks

evan2645 commented:
>The previous version tried to distinguish between identities of two pods with the same set of container images, but different counts (for instance, one pod with one foo and one bar container, and another with one foo and two bars). This is a niche scenario, and not really needed for our use case.

Feels like this might be important for security? The second `bar` could have a completely different configuration. I might be overthinking it a bit, but perhaps the `pod-image-count` selector could be `pod-container-count`? I'm still trying to think through a scenario in which this could be problematic... Haven't come up with one yet, but it feels like something is there. 🤷 food for thought

cdoron commented:
Hi @evan2645 
We could switch to a container-count (or have both container-count and image-count).
This would help us distinguish between "1 `foo`, 1 `bar`" and "1 `foo`, 2 `bar`". It would **not** distinguish between "2 `foo`, 1 `bar`" and "1 `foo`, 2 `bar`".
If we want to address such cases, we would have to go back to the previous version. In the previous version we had a counter for each image (e.g. "pod-image:foo:count=2"). This solution is a bit cumbersome, so we need to decide how we want to trade off security and elegance.

RoeeShlomo commented:
In my view we can either target set semantics or list semantics. 

Set semantics is what is currently implemented in this PR and it does provide better selectors than what's currently available. It does have the limitations described above that are inherent from the semantics of a set, but still a nice step forward.

List semantics does not have these limitations, but would require either one of the solutions described above (a signature that is hard to debug, cound per selector, or something else along those line) or proper support in the spire code for list semantics. 

Specifically, if we allowed the same selector to occur more than once and check for the number of occurrences when matching then the following is how a user would express "exactly 3 containers: foo, foo and bar":
```
pod-image:foo
pod-image:foo
pod-image:bar
pod-image-count:3
```

azdagron commented:
I agree that the set semantics that are currently in this PR are useful and a nice step forward. I've been thinking about what extra security guarantees list semantics would realistically have and I haven't been able to come up with anything. In any case, I don't think the current semantics and selector choice preclude us from doing something more intricate later.

Unless @evan2645 disagrees or has further concern, I'd be fine taking this as-is.
